x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncStatusResponse.java

```diff
@@ -0,0 +1,220 @@
+            successfulShards = searchResponse.getSuccessfulShards();
+            skippedShards = searchResponse.getSkippedShards();
+            failedShards = searchResponse.getFailedShards();
+        }
+        if (asyncSearchResponse.isRunning() == false) {
+            if (searchResponse != null) {
+                completionStatus = searchResponse.status();
+            } else {
+                Exception failure = asyncSearchResponse.getFailure();
+                if (failure != null) {
+                    completionStatus = ExceptionsHelper.status(ExceptionsHelper.unwrapCause(failure));
+                }
+            }
+        }
+        return new AsyncStatusResponse(
+            asyncSearchResponse.getId(),
+            asyncSearchResponse.isRunning(),
+            asyncSearchResponse.isPartial(),
+            asyncSearchResponse.getStartTime(),
+            expirationTimeMillis,
+            totalShards,
+            successfulShards,
+            skippedShards,
+            failedShards,
+            completionStatus
+        );
+    }
+
+    public AsyncStatusResponse(StreamInput in) throws IOException {
+        this.id = in.readString();
+        this.isRunning = in.readBoolean();
+        this.isPartial = in.readBoolean();
+        this.startTimeMillis = in.readLong();
+        this.expirationTimeMillis = in.readLong();
+        this.totalShards = in.readVInt();
+        this.successfulShards = in.readVInt();
+        this.skippedShards = in.readVInt();
+        this.failedShards = in.readVInt();
+        this.completionStatus = (this.isRunning == false) ? RestStatus.readFrom(in) : null;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(id);
+        out.writeBoolean(isRunning);
+        out.writeBoolean(isPartial);
+        out.writeLong(startTimeMillis);
+        out.writeLong(expirationTimeMillis);
+        out.writeVInt(totalShards);
+        out.writeVInt(successfulShards);
+        out.writeVInt(skippedShards);
+        out.writeVInt(failedShards);
+        if (isRunning == false) {
+            RestStatus.writeTo(out, completionStatus);
+        }
+    }
+
+    @Override
+    public RestStatus status() {
+        return OK;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field("id", id);
+        builder.field("is_running", isRunning);
+        builder.field("is_partial", isPartial);
+        builder.timeField("start_time_in_millis", "start_time", startTimeMillis);
+        builder.timeField("expiration_time_in_millis", "expiration_time", expirationTimeMillis);
+        RestActions.buildBroadcastShardsHeader(builder, params, totalShards, successfulShards, skippedShards, failedShards, null);
+        if (isRunning == false) { // completion status information is only available for a completed search
```

Comment:
I'd prefer that we always return  a  `status` even for partial responses.

Reply:
@jimczi Thanks for the comment. I am wondering how to define the status of a partial response?  
For a [completed response](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/RestStatus.java#L512-L530), we returned a failure status if we end up with `successfulShards = 0`.

For a partial response we may  have `successfulShards=0`,  simply because we haven't received yet a response from a successful shard. We can get a status only looking at the presence or absence of `failures`, but this could be a different status response from a completed search.

Reply:
Thus, I would suggest not to return a status for a partial response

Reply:
Yes the discrepancy would be weird and since we return 200 if there's at least one successful shards I agree that we should expose the status only when the search has completed.